lib/ecto/schema.ex

```diff
@@ -1991,8 +1991,10 @@ defmodule Ecto.Schema do
     pk? = opts[:primary_key] || false
     writable = opts[:writable] || :always
     put_struct_field(mod, name, Keyword.get(opts, :default))
+    redact_primary_key? = Module.get_attribute(mod, :redact) == :all_except_primary_keys and not pk?
+    redact_field? = Keyword.get(opts, :redact, false)
 
-    if Keyword.get(opts, :redact, false) do
+    if redact_primary_key? or redact_field? do
       Module.put_attribute(mod, :ecto_redact_fields, name)
     end
```

Comment:
@josevalim What do you think of this? 😄

Reply:
We only need to read the module attribute if there is no option, so the approach I originally suggested should be better suited:

```elixir
    redact_field? = Keyword.get_lazy(opts, :redact, fn ->
      Module.get_attribute(mod, :schema_redact) == :all_except_primary_keys and not pk?
    end)
```

Also, let's call the option `@schema_redact`, closer to the other module attributes, and let's add some documentation to the module docs (there is already a section on module attributes).

Reply:
I'm always blown away at how succinct your implementations are; great idea!

Updated. 😄